### PR TITLE
Flav/viz draft

### PR DESCRIPTION
**Pull request by flvndvd**

### CODE REVIEW

File: front/components/assistant/conversation/AgentMessage.tsx [hunk 1]
```diff
@@ -583,6 +556,27 @@ export function AgentMessage({
...
           </div>
         )}
+        <>
+          {agentMessage.actions
+            .filter((a) => isVisualizationActionType(a))
+            .map((a, i) => {
+              const streamingViz = streamedVisualizations.find(
+                (sv) => sv.actionId === a.id
+              );
+              assert(isVisualizationActionType(a));
```
Comment (fontanierh) [inline front/components/assistant/conversation/AgentMessage.tsx:566]: let's remove this

File: front/components/assistant/conversation/actions/VisualizationActionIframe.tsx [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  isRetryRequest,
+  isVisualizationRPCRequest,
+  visualizationExtractCodeNonStreaming,
+  visualizationExtractCodeStreaming,
+} from "@dust-tt/types";
+import { useCallback, useEffect, useState } from "react";
+
+import { RenderMessageMarkdown } from "@app/components/assistant/RenderMessageMarkdown";
+
+const sendIframeResponse = (
```
Comment (fontanierh) [inline front/components/assistant/conversation/actions/VisualizationActionIframe.tsx:19]: nit would prefer `sendResponseToIframe`

File: front/components/assistant/conversation/actions/VisualizationActionIframe.tsx [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+        sendIframeResponse(data, { file }, event.source);
+      } else if (isGetCodeToExecuteRequest(data)) {
+        const code = action.generation;
+
+        sendIframeResponse(data, { code }, event.source);
+      } else if (isRetryRequest(data)) {
+        // TODO(2024-07-24 flav) Pass the error message to the host window.
+        onRetry();
+      }
```
Comment (fontanierh) [inline front/components/assistant/conversation/actions/VisualizationActionIframe.tsx:81]: can we have an assert never ?

File: front/lib/api/assistant/actions/visualization.ts [hunk 4]
```diff
@@ -175,16 +180,61 @@ export class VisualizationConfigurationServerRunner extends BaseActionConfigurat
...
-    // Turn the conversation into a digest that can be presented to the model.
+    const contentFragmentsText = await Promise.all(
+      conversation.content
+        .flat(1)
+        .filter((m) => isContentFragmentType(m))
+        .map(async (m) => {
+          // This is needed because the ContentFragmentType is not actually narrowed correctly,
+          // this type is actually pretty complex and even if VSCode figure it out, `npm run build` does not.
+          // @flavien knows about this.
+          assert(isContentFragmentType(m));
```
Comment (fontanierh) [inline front/lib/api/assistant/actions/visualization.ts:191]: can we remove this ?

Comment (fontanierh) [reply front/lib/api/assistant/actions/visualization.ts:191]: We should use an `as` or something, it doesn't make sense to have assert here

Comment (fontanierh) [reply front/lib/api/assistant/actions/visualization.ts:191]: Don't we have this snippet logic somewhere already ?

File: front/lib/api/assistant/actions/visualization.ts [hunk 5]
```diff
@@ -175,16 +180,61 @@ export class VisualizationConfigurationServerRunner extends BaseActionConfigurat
...
+    const prompt =
+      `You have access to the following files:\n` +
+      conversation.content
+        .flat(1)
+        .filter((m) => isContentFragmentType(m))
+        .map((m, i) => {
+          // This is needed because the ContentFragmentType is not actually narrowed correctly,
+          // this type is actually pretty complex and even if VSCode figure it out, `npm run build` does not.
+          // @flavien knows about this.
+          assert(isContentFragmentType(m));
```
Comment (fontanierh) [inline front/lib/api/assistant/actions/visualization.ts:224]: Same let's remove this and use an `as`

Comment (fontanierh) [reply front/lib/api/assistant/actions/visualization.ts:224]: if we don't want an as we could remove the filter above, return null if !isContentFragmentType and use the `removeNulls` utility to get rid of all the nulls we returned (and properly narrow the type)

**Review by fontanierh — APPROVED** [review]

File: front/lib/api/assistant/actions/visualization.ts (hunk 5 above)
Comment (flvndvd) [reply front/lib/api/assistant/actions/visualization.ts:224]: Will be removed in a follow up PR.

File: front/components/assistant/conversation/AgentMessage.tsx (hunk 1 above)
Comment (flvndvd) [reply front/components/assistant/conversation/AgentMessage.tsx:566]: Will be done in a follow up PR where we clean up the types.

File: front/lib/api/assistant/actions/visualization.ts (hunk 4 above)
Comment (flvndvd) [reply front/lib/api/assistant/actions/visualization.ts:191]: We do have some snippets generated but here, it computes different snippets.

**Review by flvndvd — COMMENTED** [review]
Shipping this PR as-is so we can parallelize working on it, all comments will be addressed in follow up PRs.

**Pull request merged**